### PR TITLE
add support for patterns in scaffold decoration

### PR DESCRIPTION
**Pull request by maclandrol**
## Changelogs

- Add support for pattern-based decoration in SAFE.

---

_Checklist:_

- [ ] _Add tests to cover the fixed bug(s) or the new introduced feature(s) (if appropriate)._
- [x] _Update the API documentation if a new function is added, or an existing one is deleted. Eventually consider making a new tutorial for new features._
- [x] _Write concise and explanatory changelogs below._
- [x] _If possible, assign one of the following labels to the PR: `feature`, `fix` or `test` (or ask a maintainer to do it for you)._

---

In this PR, scaffold decoration with SAFE is extended to support molecular patterns (SMARTS). Note that the full specification of SMARTS is not supported in this implementation due to various limitations with the tokenizer and how SMARTS patterns are matched with the tokens for constrained-sampling.

### CODE REVIEW

File: safe/_pattern.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+import transformers
+
+from loguru import logger
+from tqdm.auto import tqdm
+import contextlib
+
+
+class PatternConstraint:
+    """
+    Sampling decorator for pretrained Izanagi models.
```
Comment (zhu0619) [inline safe/_pattern.py:19]: safe models

File: safe/_pattern.py [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+
+        max_dec_steps = max_length - len(self.pattern_decorator)
+        if max_dec_steps < 0:
+            raise ValueError("Step size negative due to scaffold being longer than max_length")
+
+        input_ids = start_token
+        trackers = torch.zeros(batch_size, dtype=torch.int)  # Tracks the position in the scaffold
+        torch.zeros(batch_size)
+        torch.zeros(batch_size)
+        torch.zeros(batch_size)
```
Comment (zhu0619) [inline safe/_pattern.py:446]: Are those repeated lines intended?

File: safe/_pattern.py [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+
+        [[] for _ in range(batch_size)]
```
Comment (zhu0619) [inline safe/_pattern.py:448]: same here

File: safe/_pattern.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+import numpy as np
+import safe as sf
+import torch.nn.functional as F
+import transformers
+
+from loguru import logger
+from tqdm.auto import tqdm
+import contextlib
+
+
```
Comment (zhu0619) [inline safe/_pattern.py:16]: Could you provide a few bullet points on the generation logic for both classes to make the explanation easier to understand?

File: safe/_pattern.py [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+        Initialize the PatternConstraint with a scaffold and tokenizer.
+
+        Args:
+            scaffold (str): The input scaffold for decoration.
+            tokenizer: The tokenizer to use for encoding the scaffold.
+            branch_opener (str): Character to represent branch openers.
+            branch_closer (str): Character to represent branch closers.
+            min_linker_size (int): Minimum size of linkers.
+            min_ring_size (int): Minimum size of rings.
+            force_constraint_sample (bool): Whether to force sampling constraints.
```
Comment (zhu0619) [inline safe/_pattern.py:71]: Docstring for `temperature` is missing

**Review by zhu0619 — COMMENTED** [review]
Thanks @maclandrol 

The generation process is the constrained generation based on scaffold with pattern, followed by fragment extension.
Could you also add a few bullet points docstrings on the generation logic for both classes to make the explanation easier to understand for others?

**Comment by maclandrol** [comment]
Thanks, I will address those comments.

File: safe/sample.py [hunk 6]
```diff
@@ -636,6 +637,90 @@ def scaffold_decoration(
...
+            n_samples_per_trial: Number of new molecules to generate for each randomization.
+            n_trials: Number of randomizations to perform.
+            do_not_fragment_further: Whether to prevent further fragmentation of the scaffold.
+            sanitize: Whether to sanitize the generated molecules and ensure the scaffold is present.
+            random_seed: Seed for randomization.
+            n_scaff_random: Number of scaffold randomizations to try (to reposition constraints in the string and increase rollout likelihood).
+                Increasing this will improve sampling, but will require more time.
+            n_scaff_samples: Maximum number of samples to sample for a given scaffold from the pattern.
+                Increasing this will make sure you have more diversity in the scaffold coming from the pattern
+            scaff_temperature: Temperature to use when sampling valid scaffolds from the pattern.
```
Comment (zhu0619) [inline safe/sample.py:675]: Maybe add more intuitive explanation here, something like " As the temperature increases, the generated scaffolds and structures become less confident but more diverse."

File: docs/tutorials/design-with-safe.ipynb
Comment (zhu0619) [inline docs/tutorials/design-with-safe.ipynb:1]: Could you address the error in the tutorial?  Maybe adding a type check?
```
Python argument types in
    rdkit.Chem.rdmolops.FindPotentialStereo(NoneType)
did not match C++ signature:
    FindPotentialStereo(RDKit::ROMol {lvalue} mol, bool cleanIt=False, bool flagPossible=True)
```

Comment (maclandrol) [reply docs/tutorials/design-with-safe.ipynb:1]: done

**Comment by maclandrol** [comment]
All fixed now. Thanks @zhu0619
